server/src/main/java/org/elasticsearch/index/engine/ThreadPoolMergeExecutorService.java

```diff
@@ -0,0 +1,264 @@
+        this.concurrentMergesFloorLimitForThrottling = maxConcurrentMerges * 2 + 1;
+        this.concurrentMergesCeilLimitForThrottling = maxConcurrentMerges * 4 + 1;
+    }
+
+    boolean submitMergeTask(MergeTask mergeTask) {
+        assert mergeTask.isRunning() == false;
+        // first enqueue the runnable that runs exactly one merge task (the smallest it can find)
+        if (enqueueMergeTaskExecution() == false) {
+            // if the thread pool cannot run the merge, just abort it
+            mergeTask.abortOnGoingMerge();
+            return false;
+        } else {
+            if (mergeTask.supportsIOThrottling()) {
+                // count enqueued merge tasks that support IO auto throttling, and maybe adjust IO rate for all
+                maybeUpdateIORateBytesPerSec(currentlySubmittedIOThrottledMergeTasksCount.incrementAndGet());
+            }
+            // then enqueue the merge task proper
+            queuedMergeTasks.add(mergeTask);
+            return true;
+        }
+    }
+
+    void reEnqueueBackloggedMergeTask(MergeTask mergeTask) {
+        queuedMergeTasks.add(mergeTask);
+    }
+
+    public boolean allDone() {
+        return queuedMergeTasks.isEmpty()
+            && currentlyRunningMergeTasks.isEmpty()
+            && currentlySubmittedIOThrottledMergeTasksCount.get() == 0L;
+    }
+
+    /**
+     * Enqueues a runnable that executes exactly one merge task, the smallest that is runnable at some point in time.
+     * A merge task is not runnable if its scheduler already reached the configured max-allowed concurrency level.
+     */
+    private boolean enqueueMergeTaskExecution() {
+        try {
+            executorService.execute(() -> {
+                boolean interrupted = false;
+                // one such runnable always executes a SINGLE merge task from the queue
+                // this is important for merge queue statistics, i.e. the executor's queue size represents the current amount of merges
+                MergeTask smallestMergeTask = null;
+                try {
+                    while (true) {
+                        try {
+                            // will block if there are backlogged merges until they're enqueued again
+                            smallestMergeTask = queuedMergeTasks.take();
+                            // let the task's scheduler decide if it can actually run the merge task now
+                            if (smallestMergeTask.runNowOrBacklog()) {
+                                runMergeTask(smallestMergeTask);
+                                break;
+                            }
+                            smallestMergeTask = null;
+                            // the merge task is backlogged by the merge scheduler, try to get the next smallest one
+                            // it's then the duty of the said merge scheduler to re-enqueue the backlogged merge task when it can be run
+                        } catch (InterruptedException e) {
```

Comment:
I think this can only catch interrupts from `take`, I would prefer to have the catch be local to that line - to ease readability.

I'd find it ok to use `continue` in that case - though I suspect we may only use interrupt when shutting down if at all?

Reply:
I've pushed [054fda0](https://github.com/elastic/elasticsearch/pull/120869/commits/054fda0b4a9f140c7520a8946bf11dae2f583dc7)

Reply:
>  we may only use interrupt when shutting down if at all?

Yeah, that's right, interrupt on these threads should normally be called only when shutting down the merge thread pool (i.e. shutting down the node).

Reply:
>perhaps we should not loop then but rather exit?

This is a valid option, yes. But I think there's a certain elegance to think of the merge thread pool as always executing all the enqueued merge tasks, even when the thread pool is shutting down. That means that merge tasks all always either run or aborted.
But in practice you're right that it doesn't really matter because when the thread pool shuts down, the node shuts down, so any in-memory accounting about enqueued merges are not important to be consistent when the process shuts down.

Reply:
I've NOT changed the implementation here, let me know if I should.

Reply:
I am ok with this for now. Is there adequate testing of shutdown?

Reply:
`ThreadPoolMergeSchedulerTests#testSchedulerCloseWaitsForRunningMerge` covers the case where the scheduler (not the executor service) is closing and asserts that the close method waits for the running merges to complete, and `ThreadPoolMergeExecutorServiceTests#testNewMergeTaskIsAbortedWhenThreadPoolIsShutdown` covers the case of new merges, not existing ones.

I'll follow-up with a test case where the `ThreadPoolMergeExecutorService`'s pool shut down and existing merges are left to complete.

Reply:
Pushed [b6f3f0a](https://github.com/elastic/elasticsearch/pull/120869/commits/b6f3f0a476938912fb56dcc5ea47a8598321fa10)